server/src/main.rs

```diff
@@ -39,6 +40,9 @@ fn main() {
     // current child thread (only one at a time)
     let child_thread = Arc::new(Mutex::new(None));
 
+    // (kill_flag, id of current execution)
+    let kill_flag = Arc::new((AtomicBool::new(false), AtomicIsize::new(-1))); 
```

Comment:
Maybe better than duplicating the current execution ID would be to move `Thread::execution` and `Thread::id` into their own structure with a separate lock.

(In fact, you already have a bug caused by this duplication because you don't always update `kill.1` when modifying `child.execution`.)